### DISCUSSION TITLE
onEscape, can I control the currently opened `arg()`?

### DISCUSSION THREAD

**Discussion by ramiroaraujo** [johnlindquist/kit#1353] [Q&A] [answered]
I'd like to control the `arg` when hitting escape. I have a script that performs multiple steps, and I want to hit escape and move back to the previous steps, conserving state. 
I see it should return `void` or `Promise<void>` so no option there, and I couldn't find anything useful with the `appState` param that it's passed.
I tried prompting an `arg()` inside the callback and works, but the script gets blocked (I have to manually terminate it) after I submit that one. And I can see the original `arg` (the one that triggered this new one inside `onEscape`) is never resolved.

I _think_ I can make it work by running `run()` on the same script, and passing the currently performed steps as arguments, but it feels pretty hacky, plus it looks like I loose the debugger connection if I do it that way :(

Any help/hint is appreciated
Thanks!

**Comment by ramiroaraujo** [comment, marked as answer]
I did make it work with `run()` and to my surprise, it works pretty well. It still feels hacky, since I assume it's possible (or should be?) to be done without re-starting the script over and over.

For the record, I'm passing each step argument as a JSON encoded string to the `run` command, then checking `args.length`, parse each json, and re-work the state up that that point

**Reply by ramiroaraujo** [reply]
After a careful review of `kit` source code, I can see this approach is used all the time! 😅

**Reply by johnlindquist** [reply]
@ramiroaraujo I have a "rewind" example here:

https://github.com/johnlindquist/kit/discussions/1099

Does that help answer the question?


Also, like you discovered, it's sometimes best to think of a "script" as a "function" that you may need to call recursively based on various conditions. The entire script file is a block of code that can be executed with arguments. Sounds like a function to me :)

**Reply by ramiroaraujo** [reply]
@johnlindquist yes it makes sense. The only thing I couldn't solve is to keep the debugger working. It works on the first run, but after I call `run()` it doesn't break anymore
